### ISSUE TITLE
EELS signal loss for thick samples of SrTiO<sub>3</sub>

### ISSUE THREAD

**Issue by drpaulsharp** [jacobjma/abTEM#66]
When running an EELS simulation targeting the oxygen K-edge of SrTiO<sub>3</sub> I find that, as the thickness of the sample increases, there is a loss of signal for the oxygen atoms on the titanium columns. The code and images below demonstrate the issue. After discussing this with @HamishGBrown, and using the abtem transition potentials in py_multislice, we think that the source of error may be due to a loss of information in the backward multislice step of the EELS simulation.

```python
import abtem
import ase.io
import numpy as np
import matplotlib.pyplot as plt

thicknesses = [10.0, 20.0, 50.0, 100.0, 200.0, 300.0]

probe_energy = 80e3
pixels = [64, 64]
slice_thickness = 1.9525
num_probe_positions = [64, 64]

aperture = 22.5
eels_collection_angle = 90.0
structure_tiling = [2, 2, 1]
filename = "STO_eels"

Z = 8 # atomic number
n = 1 # principal quantum number
l = 0 # azimuthal quantum number
xc = 'PBE' # exchange-correlation functional

orig_structure = ase.io.read("SrTiO3.cif") * structure_tiling


# Plot image
fig, ax = plt.subplots(nrows = 1, ncols = len(thicknesses), figsize=(5,5))

for i, thickness in enumerate(thicknesses):

    structure = orig_structure.copy()
    structure_layers = round(thickness / structure.cell.cellpar()[2])
    structure_layers = int(structure_layers)

    if structure_layers <= 0:
        structure_layers = 1

    structure = ase.build.surface(structure,
                                  indices=[0, 0, 1],
                                  layers=structure_layers,
                                  periodic=True)

    O_transitions = abtem.ionization.transitions.SubshellTransitions(Z = Z, n = n, l = l, xc = 'PBE')

    transition_potential = abtem.ionization.TransitionPotential(O_transitions,
                                                                atoms=structure,
                                                                gpts=pixels,                          
                                                                energy=probe_energy,
                                                                slice_thickness = slice_thickness)


    start_scan = np.asarray([0.0, 0.0])
    end_scan = np.asarray(structure.cell.cellpar()[:2])

    probe = abtem.waves.SMatrix(extent=(end_scan - start_scan),
                                gpts=pixels,
                                energy=probe_energy,
                                semiangle_cutoff=aperture)

    scan = abtem.scan.GridScan(start=start_scan, end=end_scan,
                               sampling=probe.ctf.nyquist_sampling)

    potential = abtem.potentials.Potential(atoms=structure,
                                           gpts=pixels,
                                           slice_thickness=slice_thickness,
                                           projection = 'infinite',
                                           parametrization = 'kirkland')
    
    detector = abtem.ionization.EELSDetector(collection_angle = eels_collection_angle,
                                             interpolation = 1)
    
    eels_measurement = probe.coreloss_scan(scan, detector, potential,
                                           transition_potential=transition_potential, pbar=False)

    eels_image = eels_measurement.interpolate(new_gpts=num_probe_positions)

    ax[i].axes.xaxis.set_ticks([])
    ax[i].axes.yaxis.set_ticks([])
    ax[i].spines['top'].set_visible(False)
    ax[i].spines['bottom'].set_visible(False)
    ax[i].spines['left'].set_visible(False)
    ax[i].spines['right'].set_visible(False)

    ax[i].imshow(np.abs(eels_image.array))

plt.show()
```

10A0:
![STO_eels_10 0](https://user-images.githubusercontent.com/44529197/145843054-158cdcf7-8944-450f-a539-caf804fdfc4a.png)

20A0:
![STO_eels_20 0](https://user-images.githubusercontent.com/44529197/145843052-febae612-f02f-4aea-85e8-53658b336e1d.png)

50A0:
![STO_eels_50 0](https://user-images.githubusercontent.com/44529197/145843047-72d028bc-5168-45a4-a8b1-575488f06fb3.png)

100A0:
![STO_eels_100 0](https://user-images.githubusercontent.com/44529197/145843044-95db8b1a-92f5-4fc5-afb8-5c6943e57885.png)

200A0:
![STO_eels_200 0](https://user-images.githubusercontent.com/44529197/145843043-06c18db7-74a5-412f-9736-09ba9f6f8ab4.png)

300A0:
![STO_eels_300 0](https://user-images.githubusercontent.com/44529197/145843038-dbc46693-a5cb-49c4-b378-05f8a07558f8.png)

**Comment by jacobjma** [comment]
Thank you for reporting. I was aware that the EELS simulations in abTEM needed more work, but it has been on the back burner for some time. I think your investigation goes a good way to help fix this. 

I need a little more time to review this and give a proper response.

**Comment by drpaulsharp** [comment]
No problem, let me know if there is anything I can do to help.

**Comment by jacobjma** [comment]
I just pushed a big update to the abTEM-dask branch including an update to PRISM-EFSTEM. The new version seems to match py_multislice in my tests, but I will do a few more tests over the next week.

I will not update the current master with this update. So if you want to test this now, you have to install the version in the dask branch. I would of course appreciate it if you can give it a test.

I also implemented EFSTEM using standard multislice and I have a test showing that the standard multislice version is identical to the PRISM version (without any approximations) within floating-point errors. 

I added three notebooks under examples/dask:
PRISM-EFSTEM: https://github.com/jacobjma/abTEM/blob/abtem-dask/examples/dask/PRISM-EFSTEM.ipynb
EFSTEM: https://github.com/jacobjma/abTEM/blob/abtem-dask/examples/dask/EFSTEM.ipynb
EFTEM: https://github.com/jacobjma/abTEM/blob/abtem-dask/examples/dask/EFTEM.ipynb

I found that the PRISM EELS algorithm is many orders of magnitude faster even for small systems. I simulated the image from the @HamishGBrown PRL paper in 7.5 minutes using my desktop with one Nvidia RTX2080 GPU. 

![fe_nanoparticle](https://user-images.githubusercontent.com/17334286/149005871-7029c520-f755-469f-8c60-77a930d9642e.png)

**Comment by drpaulsharp** [comment]
I've tested this using the PRISM algorithm today and I'm afraid that the problem I described persists. The code and images are again shown below.

During this test, I have also come across some issues in abtem.measure.images.interpolate() -- when running on a CPU rather than a GPU the routine results in a segmentation fault, and when I tried the standard multislice the result of the interpolation was an array of NaNs.

```python
import abtem
import ase.io
import numpy as np
import matplotlib.pyplot as plt

from abtem import config
config.set({'device': 'gpu'})

thicknesses = [10.0, 20.0, 50.0, 100.0, 200.0, 300.0]

probe_energy = 80e3
pixels = [64, 64]
slice_thickness = 1.9525
num_probe_positions = [64, 64]

aperture = 22.5
eels_collection_angle = 90.0
structure_tiling = [2, 2, 1]
filename = "STO_eels_prism"

Z = 8 # atomic number
n = 1 # principal quantum number
l = 0 # azimuthal quantum number
xc = 'PBE' # exchange-correlation functional

orig_structure = ase.io.read("SrTiO3.cif") * structure_tiling

for i, thickness in enumerate(thicknesses):

    fig, ax = plt.subplots(nrows = 1, ncols = 1)
    
    structure = orig_structure.copy()
    structure_layers = round(thickness / structure.cell.cellpar()[2])
    structure_layers = int(structure_layers)

    if structure_layers <= 0:
        structure_layers = 1

    structure = ase.build.surface(structure,
                                  indices=[0, 0, 1],
                                  layers=structure_layers,
                                  periodic=True)

    start_scan = np.asarray([0.0, 0.0])
    end_scan = np.asarray(structure.cell.cellpar()[:2])

    potential = abtem.potentials.Potential(atoms=structure,
                                           gpts=pixels,
                                           slice_thickness=slice_thickness,
                                           projection = 'infinite',
                                           parametrization = 'kirkland')

    O_transitions = abtem.ionization.transitions.SubshellTransitions(Z = Z, n = n, l = l, xc = 'PBE')
    transition_potential = O_transitions.get_transition_potentials(extent=(end_scan - start_scan),
                                                                          gpts=pixels,
                                                                          energy=probe_energy)

    probe = abtem.waves.SMatrix(potential,
                                planewave_cutoff=aperture,
                                extent=(end_scan - start_scan),
                                gpts=pixels,
                                energy=probe_energy)

    scan = abtem.GridScan(start=start_scan, end=end_scan)
   
    eels_measurement = probe.linear_scaling_transition_scan(scan, eels_collection_angle, transition_potential)
    eels_image = eels_measurement.interpolate(gpts=num_probe_positions)

    ax.axes.xaxis.set_ticks([])
    ax.axes.yaxis.set_ticks([])
    ax.spines['top'].set_visible(False)
    ax.spines['bottom'].set_visible(False)
    ax.spines['left'].set_visible(False)
    ax.spines['right'].set_visible(False)

    ax.imshow(np.abs(abtem.core.backend.asnumpy(eels_image.array)))
    plt.savefig(filename+"_"+str(thickness)+".png", bbox_inches='tight')
```

10A0:
![STO_eels_10 0](https://user-images.githubusercontent.com/44529197/149185910-c05dc3d7-18fd-4c30-8e52-c2d72b1b5c45.png)

20A0:
![STO_eels_20 0](https://user-images.githubusercontent.com/44529197/149185907-528f2638-9ce9-46f3-8777-563fef1fc679.png)

50A0:
![STO_eels_50 0](https://user-images.githubusercontent.com/44529197/149185904-c07ac03e-fb6e-4b16-b427-be269665079f.png)

100A0:
![STO_eels_100 0](https://user-images.githubusercontent.com/44529197/149185903-c275e977-0b14-4e4e-9f11-b05cf73ee6fd.png)

200A0:
![STO_eels_200 0](https://user-images.githubusercontent.com/44529197/149185901-5b090d07-1914-4032-9fb4-ad0537bf9d5f.png)

300A0:
![STO_eels_300 0](https://user-images.githubusercontent.com/44529197/149185896-2cbed7a5-39c0-40b4-a31d-fa39cf00baf0.png)

**Comment by jacobjma** [comment]
Hmm.. in that case, I am starting to think that it was correct all along, or at least that it agrees with py_multislice. I made some comparisons and I get almost perfect agreement for standard EFSTEM and EFTEM (using a transition potential from abTEM).

Standard EFSTEM and EFSTEM with PRISM gives identical results in abTEM. I am not able to get this agreement in py_multislice, although this could be due to a user error.

EFTEM (pyms/abTEM) (SrTiO3 - thickness = 50 Å, 20 mrad aperture, 20 mrad collection angle)
![pyms_eftem](https://user-images.githubusercontent.com/17334286/149311694-6bdc7cae-9bae-4ceb-9a08-52487259a626.png)  ![abtem_eftem](https://user-images.githubusercontent.com/17334286/149311973-245f1b06-1f81-4d2f-8f64-7d0228a65ce2.png)

EFSTEM (pyms/abTEM) (SrTiO3 - thickness = 11 Å, 20 mrad aperture, 20 mrad collection angle)
![pyms_efstem](https://user-images.githubusercontent.com/17334286/149311697-6661e6d0-3f3c-4a58-b292-a401c0fe86af.png) ![abtem_efstem](https://user-images.githubusercontent.com/17334286/149313330-02e9e1eb-001f-4ab0-8aa0-2db67456fc49.png)

EFSTEM with PRISM (pyms/abTEM) (SrTiO3 - thickness = 11 Å, 20 mrad aperture, 20 mrad collection angle)
![pyms_efstem_prism](https://user-images.githubusercontent.com/17334286/149312780-fd3f3c0d-4fd7-434d-acf6-a8b1cb83d95e.png) ![abtem_prism_efstem](https://user-images.githubusercontent.com/17334286/149313078-2f3bc164-03f0-44fa-9a9b-a97cac2db4ff.png)

I also added a `reverse_multislice` argument to the `linear_scaling_transition_scan` method, which is `False` by default. 

> During this test, I have also come across some issues in abtem.measure.images.interpolate() -- when running on a CPU rather than a GPU the routine results in a segmentation fault, and when I tried the standard multislice the result of the interpolation was an array of NaNs.

Thank you for reporting. I believe it has to do with FFTW, I will look into it.

**Comment by jacobjma** [comment]
The description of the EFTEM simulation is incorrect. An objective aperture of 60 mrad was used (and no collection angle of course).

**Comment by jacobjma** [comment]
One source of disagreement could be that "reverse multislice" was used by default before. I only get a really good agreement without reverse multislice and using it does make a significant difference for some combination of simulation parameters.

Here for a 50 Å SrTiO3:

&nbsp; &nbsp; &nbsp; &nbsp; pyms standard STEM EELS  &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; abTEM PRISM STEM EELS &nbsp; &nbsp; &nbsp; &nbsp; abTEM PRISM STEM EELS with reverse multislice

![efstem_thicker_pyms](https://user-images.githubusercontent.com/17334286/149324798-cce62b95-2cf9-49e4-a0d7-feb9bfa4dee4.png) ![abtem_prism_efstem_thicker](https://user-images.githubusercontent.com/17334286/149324878-8298b333-44b2-4a72-a25d-298e543328ad.png) ![abtem_prism_efstem_thicker_reverse](https://user-images.githubusercontent.com/17334286/149324888-30af4101-ab73-497f-87aa-a796bff646ab.png)

However, using reverse multislice does not seem to matter much for the simulation @drpaulsharp attached.

**Comment by drpaulsharp** [comment]
I've also found that the reverse multislice doesn't make much difference for my set of simulations.

What version of pyms are you using, and what set of options? When I have tried to reproduce my results in pyms I have found a strange issue where the final image is an off-centred unit cell, regardless of the tiling.

**Comment by jacobjma** [comment]
I have not had that issue you mention, but here is the code I used for simulating the STEM EELS images in py_multislice. It is using the non-dask version of abTEM for the transition potentials. I am using the latest version of py_multislice on github.

```
import torch

import os
import numpy as np
import matplotlib.pyplot as plt
from ase.io import read

from abtem.ionization import SubshellTransitions
import pyms

def export_pyms_xyz(filename, atoms):
    with open(filename, 'w') as f:
        f.write('test \n')
        f.write('    {} {} {} \n'.format(*np.diag(atoms.cell)))
        for number, position in zip(atoms.numbers, atoms.positions):
            f.write('{}  {}  {}  {}  {}  {} \n'.format(number, *position, 1, 0.00))
        f.write('-1 \n')

atoms = read('srtio3_100.cif') * (2, 2, 13)
export_pyms_xyz('srtio3_100.xyz', atoms)

Z = 8 # atomic number
n = 1 # principal quantum number
l = 0 # azimuthal quantum number
xc = 'PBE' # exchange-correlation functional

O_transitions = SubshellTransitions(Z = Z, n = n, l = l, xc = 'PBE')

transition_potentials = O_transitions.get_transition_potentials(extent = np.diag(atoms.cell)[:2], 
                                                                gpts = 256, 
                                                                energy = 80e3)

Hn0 = transition_potentials[0].build()[0]

eV = 80e3
thickness = atoms.cell[2,2]
gridshape = [256, 256]
app = 20
det = [20]
crystal = pyms.structure.fromfile('srtio3_100.xyz', atomic_coordinates='cartesian')
tiling = [1, 1]
subslices = np.linspace(0, 1, 2 * 13)
nfph = 1
PRISM_factor = [1, 1]

result = pyms.STEM_EELS_multislice(
    crystal,
    gridshape,
    eV,
    app,
    det,
    thickness,
    nfph = nfph,
    n = 0,
    ell = 0,
    epsilon = 0,
    Ztarget = 8,
    Hn0 = [Hn0],
    subslices = subslices,
    device_type = 'cpu',
    df = 0,
    tiling = tiling,
    nT = 1,
    ionization_cutoff = None,
    showProgress = 'notebook',
)
```


```

**Comment by drpaulsharp** [comment]
Thanks for this. Here's what I've found for the simulations you posted:

11Å

pyms with multislice / abTEM with multislice
![STO_pyms_multislice_11 715](https://user-images.githubusercontent.com/44529197/149798391-5d1b0e0c-41df-4d1c-9388-9c53d655231e.png)![STO_eels_multislice_11 715](https://user-images.githubusercontent.com/44529197/149798380-03237a6b-e340-428e-8d36-77a0c51ca299.png)

pyms with prism / abTEM with prism
![STO_pyms_prism_11 715](https://user-images.githubusercontent.com/44529197/149798458-6c29e767-0266-4671-a830-6bd705684674.png)![STO_eels_prism_11 715](https://user-images.githubusercontent.com/44529197/149798481-a5e5c874-f39c-44e4-b7de-305731c6470a.png)

These images seem to mostly agree with those above, but do not directly correspond. Have the images above been labelled correctly?

50Å

pyms standard STEM EELS        /           abTEM PRISM STEM EELS    /     abTEM PRISM STEM EELS with reverse multislice
![STO_pyms_50](https://user-images.githubusercontent.com/44529197/149799386-26d0fd9d-602e-4d64-ac80-a29ba8ea5551.png)![STO_eels_prism_50 0](https://user-images.githubusercontent.com/44529197/149799393-d8407e60-9630-4177-93bf-b6bc2517e704.png)![STO_eels_rm_prism_50 0](https://user-images.githubusercontent.com/44529197/149799390-870b2e36-e4b4-40f3-9b12-fd818e957712.png)

The first two images agree with those above but I find a different signal for the image using reverse multislice. I also tried comparing the use of reverse multislice for pyms with PRISM at 50Å:

pyms PRISM STEM EELS    /    pyms PRISM STEM EELS with reverse multislice
![STO_pyms_prism_50 0](https://user-images.githubusercontent.com/44529197/149801072-78bd21ed-aeac-4775-8e2c-f259ab7e321c.png)
![STO_pyms_rm_prism_50 0](https://user-images.githubusercontent.com/44529197/149801070-edc1af86-96af-44c8-b1cb-80488454cc3b.png)

It can be seen that there is much stronger agreement between the images compared to abTEM. So it looks like the reverse multislice may be causing some issues.

**Comment by jacobjma** [comment]
> These images seem to mostly agree with those above, but do not directly correspond. Have the images above been labelled correctly?

Yes, they should be correct.

> It can be seen that there is much stronger agreement between the images compared to abTEM. So it looks like the reverse multislice may be causing some issues.

Yes, I just found the error in the reverse multislice version. It is corrected now.

With this correction, I finally have good agreement across the board.

![comparison](https://user-images.githubusercontent.com/17334286/150512896-35ddd760-1bbb-4991-b71f-4bab9d2b73f0.png)

Link to notebook: https://www.dropbox.com/t/GPdTdr3ixMxY7QAK
